File: 0127-word-ladder/0127-word-ladder.py
from collections import deque
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        # word_set = set(wordList)

        # if not endWord in word_set:
        #     return 0

        # queue = deque([beginWord])

        # length = 1

        # while queue:
        #     size = len(queue)
        #     length += 1

        #     for _ in range(size):
        #         word = queue.popleft()

        #         for i in range(len(word)):
        #             for j in range(26):
        #                 next_word = word[:i] + chr(ord('a') + j) + word[i + 1:]

        #                 if next_word in word_set:
        #                     if next_word == endWord:
        #                         return length

        #                     queue.append(next_word)
        #                     word_set.remove(next_word)
        # return 0

        from collections import deque

        wordSet = set(wordList)

        if not endWord in wordSet:
            return 0

        queue = deque([beginWord])
        count = 1

        while queue:
            count += 1
            size = len(queue)

            for _ in range(size):
                word = queue.popleft()

                for idx in range(len(word)):
                    for c in ascii_lowercase:
                        newWord = word[:idx] + c + word[idx + 1:]

                        if newWord == endWord:
                            return count

                        if newWord in wordSet:
                            queue.append(newWord)
                            wordSet.remove(newWord)
        
        return 0


        # from collections import deque
        
        # wordSet = set(wordList)

        # if not endWord in wordSet:
        #     return 0

        # queue = deque([beginWord])
        # count = 0

        # while queue:
        #     size = len(queue)
        #     count += 1

        #     for _ in range(size):
        #         word = queue.popleft()

        #         for i in range(len(word)):
        #             for c in string.ascii_lowercase:
        #                 nextWord = word[:i] + c + word[i + 1:]

        #                 if nextWord in wordSet:
        #                     if nextWord == endWord:
        #                         return count + 1

        #                     queue.append(nextWord)
        #                     wordSet.remove(nextWord)

        # return 0


        

            

File: 0127-word-ladder/0127-word-ladder.swift
class Solution {
    func ladderLength(_ beginWord: String, _ endWord: String, _ wordList: [String]) -> Int {
        var wordSet = Set(wordList)

        if !wordSet.contains(endWord) {
            return 0
        }

        var queue: [String] = []
        let lowercase = "abcdefghijklmnopqrstuvwxyz"

        queue.append(beginWord)

        var count = 0

        while !queue.isEmpty {
            let size = queue.count
            count += 1

            for _ in 0..<size {
                let word = queue.removeFirst()

                for i in 0..<word.count {
                    var charArray = Array(word)
                    for c in lowercase {
                        charArray[i] = c

                        let nextWord = String(charArray)

                        if wordSet.contains(nextWord) {
                            if nextWord == endWord {
                                return count + 1
                            }
                            print(nextWord)
                            queue.append(nextWord)
                            wordSet.remove(nextWord)
                        }
                    }
                }
            }
        }

        return 0
    }
}